### PR TITLE
Add dynamic APIs to the list of discoverable APIs

### PR DESCRIPTION
**Pull request by brendandburns**
@lavalamp

### CODE REVIEW

**Comment by brendandburns** [comment]
@k8s-bot unit test this please

**Comment by brendandburns** [comment]
switching to @nikhiljindal

**Comment by brendandburns** [comment]
@nikhiljindal friendly ping.

thanks

File: pkg/master/master.go [hunk 1]
```diff
@@ -340,12 +340,17 @@ type Master struct {
 	// storage for third party objects
 	thirdPartyStorage storage.Interface
 	// map from api path to storage for those objects
```
Comment (nikhiljindal) [inline pkg/master/master.go]: Update this comment?

File: pkg/master/master.go [hunk 2]
```diff
@@ -340,12 +340,17 @@ type Master struct {
...
-	thirdPartyResources map[string]*thirdpartyresourcedataetcd.REST
+	thirdPartyResources map[string]thirdPartyEntry
 	// protects the map
 	thirdPartyResourcesLock   sync.RWMutex
 	KubernetesServiceNodePort int
 }
 
+type thirdPartyEntry struct {
```
Comment (nikhiljindal) [inline pkg/master/master.go]: Can you please add a comment for this?

**Comment by nikhiljindal** [comment]
Not sure if there is a test for discovery API but will be good to have a test that adds third party resources and then verifies that they are part of the discovery API

cc @caesarxuchao

File: pkg/master/master.go (hunk 1 above)
Comment (brendandburns) [inline pkg/master/master.go]: done.

File: pkg/master/master.go (hunk 2 above)
Comment (brendandburns) [inline pkg/master/master.go]: done.

**Comment by brendandburns** [comment]
@k8s-bot unit test this please

File: pkg/master/master_test.go [hunk 3]
```diff
@@ -343,6 +342,45 @@ func TestDiscoveryAtAPIS(t *testing.T) {
 	assert.Equal(expectGroupName, groupList.Groups[0].Name)
 	assert.Equal(expectVersions, groupList.Groups[0].Versions)
 	assert.Equal(expectPreferredVersion, groupList.Groups[0].PreferredVersion)
+
+	master.thirdPartyResources["/apis/company.com/v1"] = thirdPartyEntry{
```
Comment (caesarxuchao) [inline pkg/master/master_test.go]: @nikhiljindal, I think this is the test you'd like to add.

**Comment by brendandburns** [comment]
Yep, @nikhiljindal I added the test and addressed comments, ptal.

thanks!
--brendan

File: pkg/master/master_test.go [hunk 4]
```diff
@@ -343,6 +342,45 @@ func TestDiscoveryAtAPIS(t *testing.T) {
...
+		nil,
+		unversioned.APIGroup{
+			Name: "company.com",
+			Versions: []unversioned.GroupVersionForDiscovery{
+				{GroupVersion: "company.com/v1", Version: "v1"},
+			},
+			PreferredVersion: unversioned.GroupVersionForDiscovery{GroupVersion: "company.com/v1", Version: "v1"},
```
Comment (nikhiljindal) [inline pkg/master/master_test.go]: You are initializing the same unversioned.GroupVersionForDiscovery 4 times. You can initialize it once and reuse it at other places?

File: pkg/master/master_test.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/master/master_test.go]: Yes! Thanks @brendandburns for adding it.

**Comment by nikhiljindal** [comment]
LGTM, thanks!

File: pkg/master/master_test.go (hunk 4 above)
Comment (brendandburns) [inline pkg/master/master_test.go]: done.

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

**Pull request merged**